src/browser/cached_resource_fetcher.ts

```diff
@@ -13,85 +13,213 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+            nativeImage.createFromBuffer(new Buffer(this.data, 'binary')) :
+            nativeImage.createFromDataURL(getDataURL('blank-1x1.png'));
+        return Promise.resolve(this.image);
+    }
+
+    // used to createFromPath from an .ico file that is already a file
+    public getNativeImageFromPath(path: string): Promise<NativeImage> {
+        this.image = this.image || this.success && isICO(path) ?
+            nativeImage.createFromPath(path) :
+            FetchResponse.prototype.getNativeImage.call(this);
+        return Promise.resolve(this.image);
+    }
+
+    // used to createFromPath from an .ico file created from downloaded data
+    public getNativeImageFromCreatedPath(url: string): Promise<NativeImage> {
+        return this.image = this.image || this.success && isICO(url) ?
+            makeAppCacheDir('ico')
+                .then(dirPath => getFile(dirPath, url, this.data, 'binary'))
+                .then(filePath => createFromPath(filePath, 20000)) : // 20s = small window for possible reuse
+            Promise.resolve(FetchResponse.prototype.getNativeImage.call(this));
+    }
+}
+
+type Fetcher = (url: string, encoding: string) => Promise<FetchResponse>;
 
 /**
- * Downloads a file if it doesn't exist in cache yet.
+ * Downloads a file to cache and/or retrieves it from cache and returns its status code, headers, and data
  */
-export function cachedFetch(appUuid: string, fileUrl: string, callback: (error: null|Error, path?: string) => any): void {
-    if (!fileUrl || typeof fileUrl !== 'string') {
-        callback(new Error(`Bad file url: '${fileUrl}'`));
-        return;
+export function cachedFetch(url: string, encoding: string = 'utf8'): Promise<FetchResponse> {
+    if (!url || typeof url !== 'string') {
+        return Promise.reject(new Error(`Bad file url: '${url}'`));
     }
-    if (appQuiting) {
-        callback(new Error('Runtime is exiting'));
-        return;
+
+    if (appQuitting) {
+        return Promise.reject(new Error(ERR_QUITTING));
     }
 
-    if (!isURL(fileUrl)) {
-        if (isURI(fileUrl)) {
-            callback(null, uriToPath(fileUrl));
-        } else {
-            stat(fileUrl, (err: null|Error) => {
-                if (err) {
-                    callback(new Error(`Invalid file url: '${fileUrl}'`));
+    const fetcher: Fetcher = isURL(url) ? netRequester : fileRequester;
+
+    return fetcher(url, encoding);
+}
+
+function fileRequester(url: string, encoding: string): Promise<FetchResponse> {
+    return new Promise((resolve, reject) => {
+        // remove possible URI (file:/// scheme) prefix
+        const filepath: string = url.replace(Patterns.URI, '');
+        const isImage = isImageFile(url) || isICO(url);
+
+        if (isImage) {
+            encoding = 'binary';
+        }
+
+        readFile(filepath, encoding, (error: Error | string, data: string) => {
+            if (error) {
+                if (/ENOENT/.test((<Error>error).message || <string>error)) {
```

Comment:
whats the difference between this and the reject case?

Reply:
This (file-not-found) is not considered an error; it is a fetch with a falsy `success`; all other errors are transport errors and need to be logged.